### PR TITLE
Fix serdes cache by coordinates

### PR DESCRIPTION
**Pull request by carlesarnal**

### CODE REVIEW

**Review by jsenko — CHANGES_REQUESTED** [review]
This looks good, but I think at this point it's not useful to have both
1. `schemaCache.getByArtifactReference` and
2. `schemaCache.getByArtifactCoordinates` methods.

Looking at the usages of (1), we could easily remove it and use the other methods. You're basically resolving by GAV coordinates anyway. The downside of this is, that `ArtifactReference` would have to be converted into `ArtifactCoordinates`. Also, `ArtifactReference` does not contain only GAV, but potentially also GID and CID, which would require checking and using some of the other methods, like `getByGlobalId`. Can actually `artifactResolverStrategy.artifactReference` return an ArtifactReference that does not use GAV, but GID or CID?

Alternatively, replace (2) by still having `ArtifactCoordinates` as a key in one of the cache indexes, but not exposing it outside the cache, and just use it internally within the `getByArtifactReference` method.

Either way, it is not useful to have `ArtifactReference` index in the cache, when the other indexes already contain all mappings: (G, A, V) (`ArtifactCoordinates`) and GID and CID.

WDYT?

File: schema-resolver/src/main/java/io/apicurio/registry/resolver/strategy/ArtifactCoordinates.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+        result = 31 * result + (artifactId != null ? artifactId.hashCode() : 0);
+        result = 31 * result + (version != null ? version.hashCode() : 0);
+        return result;
+    }
+
+    /**
+     * @see io.apicurio.registry.resolver.strategy.ArtifactReference#equals(java.lang.Object)
+     */
+    @Override
+    public boolean equals(Object obj) {
```
[the remark this answers is not in the document]
Comment (carlesarnal) [reply schema-resolver/src/main/java/io/apicurio/registry/resolver/strategy/ArtifactCoordinates.java:87]: Right.

**Comment by carlesarnal** [comment]
> This looks good, but I think at this point it's not useful to have both
> 
>     1. `schemaCache.getByArtifactReference` and
> 
>     2. `schemaCache.getByArtifactCoordinates` methods.
> 
> 
> Looking at the usages of (1), we could easily remove it and use the other methods. You're basically resolving by GAV coordinates anyway. The downside of this is, that `ArtifactReference` would have to be converted into `ArtifactCoordinates`. Also, `ArtifactReference` does not contain only GAV, but potentially also GID and CID, which would require checking and using some of the other methods, like `getByGlobalId`. Can actually `artifactResolverStrategy.artifactReference` return an ArtifactReference that does not use GAV, but GID or CID?
> 
> Alternatively, replace (2) by still having `ArtifactCoordinates` as a key in one of the cache indexes, but not exposing it outside the cache, and just use it internally within the `getByArtifactReference` method.
> 
> Either way, it is not useful to have `ArtifactReference` index in the cache, when the other indexes already contain all mappings: (G, A, V) (`ArtifactCoordinates`) and GID and CID.
> 
> WDYT?

Agreed, I have removed the unnecessary method and added some checks to make the cache work a bit better. Now, if a version is specified when serializing, the schema will be correctly returned from the cache.

That said, there is still the question of what to do when the version is null (return the latest schema present in the cache? Check the cache by content? (with the downside in performance, since the schema must be parsed for that)), since this will only fix the situation with a non-null version.

**Comment by jsenko** [comment]
no objection

**Comment by carlesarnal** [comment]
OK, then please, approve and remove the requested changes (I would like to avoid force merging), thanks!

**Review by jsenko — APPROVED** [review]

**Pull request merged**